File: sdks/ts/memory/src/ingest/dead-letter.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Dead letter queue for permanent failure isolation. Jobs that exhaust
 * their retry budget are moved here with full error context so that
 * operators can inspect, retry, or purge them.
 *
 * Provides an in-memory adapter for testing/local use and a type-safe
 * interface that PostgreSQL and SQLite adapters implement.
 */

import { randomUUID } from 'node:crypto'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Serialised job data preserved when a job is moved to the DLQ. */
export type JobPayload = {
  readonly documentHash: string
  readonly brainId: string
  readonly source?: string
  readonly contentType?: string
}

/** A job that has been moved to the dead letter queue. */
export type DeadLetterEntry = {
  readonly id: string
  readonly originalJobId: string
  readonly brainId: string
  readonly payload: JobPayload
  readonly failureReason: string
  readonly lastError: string
  readonly errorHistory?: readonly string[]
  readonly retryCount: number
  readonly metadata?: Readonly<Record<string, string>>
  readonly groupId?: string
  readonly movedAt: Date
  readonly resolvedAt?: Date
  readonly resolvedBy?: string
}

/** Options for listing dead letter entries. */
export type DeadLetterListOptions = {
  readonly brainId?: string
  readonly limit?: number
  readonly offset?: number
  readonly includeResolved?: boolean
}

/** Paginated result from a list operation. */
export type DeadLetterListResult = {
  readonly entries: readonly DeadLetterEntry[]
  readonly total: number
}

/** Discriminated union for purge operations. */
export type PurgeOptions =
  | { readonly kind: 'by-id'; readonly id: string }
  | { readonly kind: 'by-brain'; readonly brainId: string }
  | { readonly kind: 'older-than'; readonly days: number }
  | { readonly kind: 'all-resolved' }

/**
 * Callback that creates a new queue job in the main queue from a dead
 * letter entry. The implementation is responsible for resetting the
 * retry count and assigning a new job ID. Returning a rejected promise
 * prevents the DLQ entry from being marked as resolved.
 */
export type ReEnqueueFn = (entry: DeadLetterEntry) => Promise<void>

/** Interface for dead letter queue operations. */
export type DeadLetterAdapter = {
  /** Move a failed job to the dead letter queue. */
  readonly move: (entry: DeadLetterEntry) => Promise<DeadLetterEntry>

  /** List dead letter entries with pagination and filtering. */
  readonly list: (opts?: DeadLetterListOptions) => Promise<DeadLetterListResult>

  /** Get a single entry by ID. Returns undefined when not found. */
  readonly get: (id: string) => Promise<DeadLetterEntry | undefined>

  /**
   * Mark an entry as resolved for retry. If reEnqueue is provided, it
   * is called with the entry to create a new queue job before the
   * entry is committed as resolved. If reEnqueue throws, the entry
   * remains unresolved.
   */
  readonly retry: (id: string, resolvedBy?: string, reEnqueue?: ReEnqueueFn) => Promise<DeadLetterEntry>

  /** Purge entries matching the given options. Returns count removed. */
  readonly purge: (opts: PurgeOptions) => Promise<number>

  /** Count unresolved entries, optionally filtered by brain ID. */
  readonly count: (brainId?: string) => Promise<number>
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

/** Dead letter entry not found. */
export class DeadLetterNotFoundError extends Error {
  constructor(id: string) {
    super(`Dead letter entry not found: ${id}`)
    this.name = 'DeadLetterNotFoundError'
  }
}

/** Dead letter entry already resolved (double retry). */
export class DeadLetterAlreadyResolvedError extends Error {
  constructor(id: string) {
    super(`Dead letter entry already resolved: ${id}`)
    this.name = 'DeadLetterAlreadyResolvedError'
  }
}

// ---------------------------------------------------------------------------
// Default configuration
// ---------------------------------------------------------------------------

/** Default maximum failed attempts before a job is moved to the DLQ. */
export const DEFAULT_MAX_ATTEMPTS = 3

/** Default retention period in days for dead letter entries. */
export const DEFAULT_RETENTION_DAYS = 30

/** Default page size for list operations. */
const DEFAULT_LIST_LIMIT = 50

// ---------------------------------------------------------------------------
// In-memory adapter (testing / local use)
// ---------------------------------------------------------------------------

/**
 * Creates an in-memory dead letter adapter. Useful for testing and
 * local/embedded deployments that do not need persistent storage.
 */
export const createInMemoryDeadLetterAdapter = (): DeadLetterAdapter => {
  const store = new Map<string, DeadLetterEntry>()

  const move = async (entry: DeadLetterEntry): Promise<DeadLetterEntry> => {
    const resolved: DeadLetterEntry = {
      ...entry,
      id: entry.id || randomUUID(),
      movedAt: entry.movedAt ?? new Date(),
    }
    store.set(resolved.id, resolved)
    return resolved
  }

  const list = async (opts?: DeadLetterListOptions): Promise<DeadLetterListResult> => {
    const limit = opts?.limit ?? DEFAULT_LIST_LIMIT
    const offset = opts?.offset ?? 0
    const includeResolved = opts?.includeResolved ?? false
    const brainId = opts?.brainId

    const filtered = Array.from(store.values()).filter((entry) => {
      if (!includeResolved && entry.resolvedAt !== undefined) return false
      if (brainId !== undefined && entry.brainId !== brainId) return false
      return true
    })

    // Sort by movedAt descending (most recent first).
    filtered.sort((a, b) => b.movedAt.getTime() - a.movedAt.getTime())

    const total = filtered.length
    const entries = filtered.slice(offset, offset + limit)
    return { entries, total }
  }

  const get = async (id: string): Promise<DeadLetterEntry | undefined> => {
    return store.get(id)
  }

  const retry = async (id: string, resolvedBy?: string, reEnqueue?: ReEnqueueFn): Promise<DeadLetterEntry> => {
    const entry = store.get(id)
    if (entry === undefined) {
      throw new DeadLetterNotFoundError(id)
    }
    if (entry.resolvedAt !== undefined) {
      throw new DeadLetterAlreadyResolvedError(id)
    }

    const resolved: DeadLetterEntry = {
      ...entry,
      resolvedAt: new Date(),
      resolvedBy: resolvedBy ?? 'system',
    }

    if (reEnqueue !== undefined) {
      await reEnqueue(resolved)
    }

    store.set(id, resolved)
    return resolved
  }

  const purge = async (opts: PurgeOptions): Promise<number> => {
    const purgeStrategies: Record<PurgeOptions['kind'], () => number> = {
      'by-id': () => {
        const optsById = opts as { readonly kind: 'by-id'; readonly id: string }
        return store.delete(optsById.id) ? 1 : 0
      },
      'by-brain': () => {
        const optsByBrain = opts as { readonly kind: 'by-brain'; readonly brainId: string }
        let removed = 0
        for (const [id, entry] of store) {
          if (entry.brainId === optsByBrain.brainId) {
            store.delete(id)
            removed++
          }
        }
        return removed
      },
      'older-than': () => {
        const optsOlder = opts as { readonly kind: 'older-than'; readonly days: number }
        const cutoff = new Date()
        cutoff.setDate(cutoff.getDate() - optsOlder.days)
        let removed = 0
        for (const [id, entry] of store) {
          if (entry.movedAt < cutoff) {
            store.delete(id)
            removed++
          }
        }
        return removed
      },
      'all-resolved': () => {
        let removed = 0
        for (const [id, entry] of store) {
          if (entry.resolvedAt !== undefined) {
            store.delete(id)
            removed++
          }
        }
        return removed
      },
    }

    return purgeStrategies[opts.kind]()
  }

  const count = async (brainId?: string): Promise<number> => {
    let total = 0
    for (const entry of store.values()) {
      if (entry.resolvedAt !== undefined) continue
      if (brainId !== undefined && entry.brainId !== brainId) continue
      total++
    }
    return total
  }

  return { move, list, get, retry, purge, count } as const
}

File: go/ingest/deadletter.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"
)

// DefaultMaxAttempts is the default number of failed attempts before a job
// is moved to the dead letter queue.
const DefaultMaxAttempts = 3

// DefaultRetentionDays is the default number of days dead letter entries
// are retained before they become eligible for age-based purging.
const DefaultRetentionDays = 30

// ErrDeadLetterNotFound signals that the requested dead letter entry does
// not exist.
var ErrDeadLetterNotFound = errors.New("ingest: dead letter entry not found")

// ErrDeadLetterAlreadyResolved signals that the entry has already been
// retried or otherwise resolved.
var ErrDeadLetterAlreadyResolved = errors.New("ingest: dead letter entry already resolved")

// JobPayload holds the serialised job data that was being processed when
// the job failed permanently. This is the data needed to re-enqueue the
// job for a retry from the dead letter queue.
type JobPayload struct {
	DocumentHash string `json:"documentHash"`
	BrainID      string `json:"brainId"`
	Source       string `json:"source,omitempty"`
	ContentType  string `json:"contentType,omitempty"`
}

// DeadLetterEntry represents a job that has been moved to the dead letter
// queue after exhausting its retry budget. It preserves the full error
// history and original job context for operator inspection.
type DeadLetterEntry struct {
	ID            string            `json:"id"`
	OriginalJobID string            `json:"originalJobId"`
	BrainID       string            `json:"brainId"`
	Payload       JobPayload        `json:"payload"`
	FailureReason string            `json:"failureReason"`
	LastError     string            `json:"lastError"`
	ErrorHistory  []string          `json:"errorHistory,omitempty"`
	RetryCount    int               `json:"retryCount"`
	Metadata      map[string]string `json:"metadata,omitempty"`
	GroupID       string            `json:"groupId,omitempty"`
	MovedAt       time.Time         `json:"movedAt"`
	ResolvedAt    *time.Time        `json:"resolvedAt,omitempty"`
	ResolvedBy    string            `json:"resolvedBy,omitempty"`
}

// DeadLetterListOptions controls filtering and pagination when listing
// dead letter entries.
type DeadLetterListOptions struct {
	BrainID         string
	Limit           int
	Offset          int
	IncludeResolved bool
}

// DeadLetterListResult holds a page of dead letter entries and the total
// count matching the filter.
type DeadLetterListResult struct {
	Entries []DeadLetterEntry
	Total   int
}

// PurgeKind discriminates the type of purge operation.
type PurgeKind string

const (
	PurgeByID        PurgeKind = "by-id"
	PurgeByBrain     PurgeKind = "by-brain"
	PurgeOlderThan   PurgeKind = "older-than"
	PurgeAllResolved PurgeKind = "all-resolved"
)

// PurgeOptions specifies which dead letter entries should be removed.
type PurgeOptions struct {
	Kind    PurgeKind
	ID      string
	BrainID string
	Days    int
}

// ReEnqueueFunc is a callback that creates a new queue job in the main
// queue from a dead letter entry. The implementation is responsible for
// resetting the retry count and assigning a new job ID. Returning an
// error prevents the DLQ entry from being marked as resolved.
type ReEnqueueFunc func(ctx context.Context, entry DeadLetterEntry) error

// DeadLetterAdapter defines the operations available on the dead letter
// queue. Both SQLite and PostgreSQL implementations satisfy this
// interface.
type DeadLetterAdapter interface {
	// Move transfers a failed job to the dead letter queue with a reason.
	Move(ctx context.Context, entry DeadLetterEntry) (DeadLetterEntry, error)

	// List returns a paginated, filtered view of dead letter entries.
	List(ctx context.Context, opts DeadLetterListOptions) (DeadLetterListResult, error)

	// Get retrieves a single entry by ID. Returns ErrDeadLetterNotFound
	// when no entry exists.
	Get(ctx context.Context, id string) (DeadLetterEntry, error)

	// Retry marks an entry as resolved and invokes the re-enqueue
	// callback to create a new queue job in the main queue. Returns
	// ErrDeadLetterAlreadyResolved if the entry was already retried.
	// If reEnqueue is nil, the entry is marked as resolved without
	// re-enqueueing (caller takes responsibility).
	Retry(ctx context.Context, id string, resolvedBy string, reEnqueue ReEnqueueFunc) (DeadLetterEntry, error)

	// Purge removes entries matching the given options and returns the
	// count of entries removed.
	Purge(ctx context.Context, opts PurgeOptions) (int, error)

	// Count returns the number of dead letter entries, optionally
	// filtered by brain ID.
	Count(ctx context.Context, brainID string) (int, error)
}

// deadLetterListCap is the initial slice capacity for list results.
const deadLetterListCap = 32

// deadLetterColumns is the column list used in SELECT queries for dead
// letter entries. Both adapters reference the same columns.
const deadLetterColumns = `id, original_job_id, brain_id, payload, failure_reason, last_error,
	       error_history, retry_count, metadata, group_id, moved_at, resolved_at, resolved_by`

// --- Shared helpers ----------------------------------------------------------

func nullableString(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func nullableBytes(b []byte) *string {
	if len(b) == 0 {
		return nil
	}
	s := string(b)
	return &s
}

func nullableTime(t *time.Time) *string {
	if t == nil {
		return nil
	}
	s := t.Format(time.RFC3339)
	return &s
}

// scanner is the common interface satisfied by both *sql.Row and *sql.Rows,
// allowing a single scan function to handle both cases per dialect.
type scanner interface {
	Scan(dest ...any) error
}

// populateCommonFields deserialises JSON fields and assigns nullable
// columns that are common to both SQLite and PostgreSQL scan paths.
func populateCommonFields(
	entry *DeadLetterEntry,
	payloadJSON string,
	lastError sql.NullString,
	errorHistoryJSON sql.NullString,
	metadataJSON sql.NullString,
	groupID sql.NullString,
	resolvedBy sql.NullString,
) error {
	if err := json.Unmarshal([]byte(payloadJSON), &entry.Payload); err != nil {
		return fmt.Errorf("ingest: unmarshal dead letter payload: %w", err)
	}

	if lastError.Valid {
		entry.LastError = lastError.String
	}

	if errorHistoryJSON.Valid && errorHistoryJSON.String != "" {
		var history []string
		if unmarshalErr := json.Unmarshal([]byte(errorHistoryJSON.String), &history); unmarshalErr != nil {
			return fmt.Errorf("ingest: unmarshal dead letter error history: %w", unmarshalErr)
		}
		entry.ErrorHistory = history
	}

	if metadataJSON.Valid {
		meta := make(map[string]string)
		if unmarshalErr := json.Unmarshal([]byte(metadataJSON.String), &meta); unmarshalErr != nil {
			return fmt.Errorf("ingest: unmarshal dead letter metadata: %w", unmarshalErr)
		}
		entry.Metadata = meta
	}

	if groupID.Valid {
		entry.GroupID = groupID.String
	}
	if resolvedBy.Valid {
		entry.ResolvedBy = resolvedBy.String
	}

	return nil
}

// buildListWhere constructs a WHERE clause and parameter list for the
// list query. The placeholder argument controls the style: use "?" for
// SQLite or "$" for PostgreSQL (numbered placeholders).
func buildListWhere(opts DeadLetterListOptions, placeholder string) (string, []any) {
	clauses := make([]string, 0, 2)
	args := make([]any, 0, 2)
	paramIdx := 1

	if opts.BrainID != "" {
		ph := "?"
		if placeholder == "$" {
			ph = fmt.Sprintf("$%d", paramIdx)
		}
		clauses = append(clauses, "brain_id = "+ph)
		args = append(args, opts.BrainID)
		paramIdx++
	}
	if !opts.IncludeResolved {
		clauses = append(clauses, "resolved_at IS NULL")
	}

	_ = paramIdx
	if len(clauses) == 0 {
		return "", nil
	}
	return "WHERE " + strings.Join(clauses, " AND "), args
}

// Compile-time interface assertions.
var _ DeadLetterAdapter = (*SqliteDeadLetterAdapter)(nil)
var _ DeadLetterAdapter = (*PostgresDeadLetterAdapter)(nil)
